torch/csrc/jit/autodiff.cpp

```diff
@@ -34,38 +45,233 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
   return fmap(sym_grads, [](const SymbolicVariable &v) { return v.value(); });
 }
 
-void differentiate(std::shared_ptr<Graph>& graph) {
-  JIT_ASSERT(graph->stage() == 0);
-  graph->advanceStage();
+// Before:
+//   - graph has only stage 0
+//   - grad_desc doesn't have any fields initialized
+// After:
+//   - graph has stage 0 and stage 1 that computes its vjp
+//   - grad_desc has df_input_vjps and df_output_vjps set
+//     (but df_input_vjps will be modified later as well)
+static value_map addReverseInline(Graph& graph, Gradient& grad_desc) {
+  JIT_ASSERT(graph.stage() == 0);
```

Comment:
Is stage() == 0 a fundamental limitation or simply "we haven't tested with stage() == 1, so dunno if it works or not"?

Reply:
it's stage() == 0, because that's the regime it will be used in right now, and I don't want to generalize the semantics to arbitrary stages, because that will only make the code more complicated.